src/cgutils.cpp

```diff
@@ -1426,26 +1423,34 @@ static void typed_store(jl_codectx_t &ctx,
     if (type_is_ghost(elty))
         return;
     Value *r;
-    if (!isboxed) {
+    if (!isboxed)
         r = emit_unbox(ctx, elty, rhs, jltype);
-        if (parent != NULL)
-            emit_write_multibarrier(ctx, parent, r);
-    }
-    else {
+    else
         r = maybe_decay_untracked(boxed(ctx, rhs));
-        if (parent != NULL)
-            emit_write_barrier(ctx, parent, r);
-    }
     Type *ptrty = PointerType::get(elty, ptr->getType()->getPointerAddressSpace());
     if (ptr->getType() != ptrty)
         ptr = ctx.builder.CreateBitCast(ptr, ptrty);
     if (idx_0based)
         ptr = ctx.builder.CreateInBoundsGEP(r->getType(), ptr, idx_0based);
-    Instruction *store = ctx.builder.CreateAlignedStore(r, ptr, isboxed || alignment ? alignment : julia_alignment(jltype));
+    if (!isboxed && CountTrackedPointers(elty).count)
+        ctx.builder.CreateFence(AtomicOrdering::Release);
```

Comment:
IIUC, I don't think the fence can replace atomic here. (also more places in C code). It "improves" the ordering but not atomicity.

Reply:
I think you're confusing it with sequential consistency fences—that cannot be adequately defined as a fence, since it needs to be a property of the load/store operation itself. The release fence is documented to generally be okay here, and in practice is always okay on hardware.

Reply:
Well, what I said is that as long as the store is not atomic LLVM is allowed to do whatever it want including making the store non-atomic.